ddtrace/internal/datadog/profiling/stack_v2/src/sampler.cpp

```diff
@@ -56,17 +58,63 @@ Sampler::get()
     return instance;
 }
 
+void
+_stack_v2_atfork_child()
+{
+    // The only thing we need to do at fork is to propagate the PID to echion
+    // so we don't even reveal this function to the user
+    _set_pid(getpid());
+}
+
+__attribute__((constructor)) void
+_stack_v2_init()
+{
+    _stack_v2_atfork_child();
+}
+
 void
 Sampler::one_time_setup()
 {
     _set_cpu(true);
     init_frame_cache(echion_frame_cache_size);
-    _set_pid(getpid());
+
+    // It is unlikely, but possible, that the caller has forked since application startup, but before starting echion.
+    // Run the atfork handler to ensure that we're tracking the correct process
+    _stack_v2_atfork_child();
+    pthread_atfork(nullptr, nullptr, _stack_v2_atfork_child);
 
     // Register our rendering callbacks with echion's Renderer singleton
     Renderer::get().set_renderer(renderer_ptr);
 }
 
+void
+Sampler::register_thread(uintptr_t id, uint64_t native_id, const char* name)
```

Comment:
When I look at echion's `track_thread`
https://github.com/sanchda/echion/blob/b2f2d49f2f5d5c4dd78d1d9b83280db8ac2949c0/echion/coremodule.cc#L278-L288

It updates thread info if it's already tracked. Don't we need to do the same?

Reply:
Unless I'm mistaken, I believe we do.
1. Check that the iterator is the end iterator.  If so, then we didn't find the thing, so add it.
2. Otherwise, the iterator found a matching element, so just update the element (`it->second = std::make_unique<...>(...)...`)

Reply:
Oh, I have neglected that part. Thanks for pointing it out.